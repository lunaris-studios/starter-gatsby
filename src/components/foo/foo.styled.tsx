import * as React from "react"
import * as SC from "styled-components"

// re-import `styled-components` development mode DOM classnames.
import styled, { css } from "styled-components"

/**
 * Table of Contents
 *
 * [Foo]
 */

/**
 * [Foo]
 * - [Foo.Container(WRAPPER)]
 */

interface Foo {
  Container: SC.StyledComponent<
    "div",
    any,
    IFooContainerAttrs,
    keyof IFooContainerAttrs
  >
}

export const Foo = {} as Foo

/**
 * [Foo.Container]
 */

interface IFooContainerProps extends SC.ThemeProps<SC.DefaultTheme> {}

interface IFooContainerAttrs extends IFooContainerProps {}

Foo.Container = styled("div").attrs(
  (props: IFooContainerProps): IFooContainerAttrs => ({
    ...props,
  })
)`
	// ...
`
