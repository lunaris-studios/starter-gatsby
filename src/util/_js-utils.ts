import * as Common from "~/common"

/**
 * Returns the difference in length between two arrays. A `null` argument is
 * considered an empty list. The return value will be positive if `a` is longer
 * than `b`, negative if the opposite is true, and zero if their lengths are
 * equal.
 */
export function arrayLengthCompare(a: any[] = [], b: any[] = []) {
    return a.length - b.length;
}

/**
 * Returns true if the two numbers are within the given tolerance of each other.
 * This is useful to correct for floating point precision issues, less useful
 * for integers.
 */
export function approxEqual(a: number, b: number, tolerance = 0.00001) {
    return Math.abs(a - b) <= tolerance;
}

/** Returns the number of decimal places in the given number. */
export function countDecimalPlaces(num: number) {
    if (!isFinite(num)) {
        return 0;
    }
    let e = 1;
    let p = 0;
    while (Math.round(num * e) / e !== num) {
        e *= 10;
        p++;
    }
    return p;
}
