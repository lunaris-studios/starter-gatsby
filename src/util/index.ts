export * from "./_function-utils";
export * from "./_js-utils";
export * from "./_react-utils";

export { default as isEqual } from "lodash/isEqual";
