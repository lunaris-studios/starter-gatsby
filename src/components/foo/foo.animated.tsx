import * as Spring from "react-spring";

import * as Common from "~/common";
import * as Util from "~/util";

import * as Component from "./foo";

export enum Type {
	CONTAINER_TRANSITION = "CONTAINER_TRANSITION",
}

export interface IContainerTransition {
	opacity: number;
	transform: string;
}

export interface IContainerTransitionOptions extends Component.IFooProps {}

const defaultContainerTransition = Object.freeze<IContainerTransition>({
	opacity: 0,
	transform: "translate3d(0%, -100%, 0%)",
});

export class Container {
	public controller: Spring.Controller = new Spring.Controller();

	public static defaultContainerTransition: IContainerTransition = defaultContainerTransition;

	public transitionProps(options: IContainerTransitionOptions) {
		const { foo } = options;

		const props: Spring.TransitionDefaultProps = Object.freeze({
			from: defaultContainerTransition,
			enter: () => async (next) => {
				await next({
					opacity: 1,
					transform: "translate3d(100%, 0%, 0%)",
				});
			},
			leave: () => async (next) => {
				await next({
					opacity: 0,
					transform: "translate3d(100%, 0%, 0%)",
				});
			},
			onStart: () => {},
			onRest: () => {},
		});

		return props;
	}
}
