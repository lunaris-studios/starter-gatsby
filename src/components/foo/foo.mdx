---
title: Foo
component: Foo
description: Foo implements Bar
---

import { Foo } from "./foo.example";