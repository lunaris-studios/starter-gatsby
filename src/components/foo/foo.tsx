import * as React from "react";
import * as Spring from "react-spring";

import * as Common from "~/common";
import * as Components from "~/components";

import * as Animated from "./foo.animated";
import * as Styled from "./foo.styled";

/**
 * [Foo]
 * - [Foo.Container(WRAPPER)]
 */

export interface IFooProps {
	/**
	 * Indicates whether or not the foo is foo
	 * @default true
	 */
	foo: boolean;
}

export interface IFooState {
	/**
	 * Indicates whether or not the foo is bar
	 * @default false
	 */
	bar: boolean;
}

export interface IFooAnimated {
	/**
	 * Transition animation for [Foo.Container]
	 */
	container: Animated.Container;
}

const defaultProps = Object.freeze<IFooProps>({
	foo: true,
});

const defaultState = Object.freeze<IFooState>({
	bar: false,
});

const defaultAnimated = Object.freeze<IFooAnimated>({
	container: new Animated.Container(),
});

export class Foo extends React.PureComponent<IFooProps, IFooState, {}> {
	public static readonly displayName = `${Common.DISPLAYNAME_PREFIX}.Overlay`;

	public static readonly defaultProps: IFooProps = defaultProps;
	public static readonly defaultState: IFooState = defaultState;
	public static readonly defaultAnimated: IFooAnimated = defaultAnimated;

	public state: IFooState = defaultState;
	public animated: IFooAnimated = defaultAnimated;

	public render() {
		const {
			foo,
			// ...
		} = this.props;

		const {
			bar,
			// ...
		} = this.state;

		const containerTransitionProps = this.animated.container.transitionProps({
			foo,
		});

		return (
			<Spring.Transition items={foo} {...containerTransitionProps}>
				{(style, visible) =>
					visible && <Styled.Foo.Container style={style}>{/* ... */}</Styled.Foo.Container>
				}
			</Spring.Transition>
		);
	}

	public componentDidMount() {
		// ...
	}

	public componentDidUpdate(prevProps: IFooProps) {
		// ...
	}

	public componentWillUnmount() {
		// ...
	}
}
