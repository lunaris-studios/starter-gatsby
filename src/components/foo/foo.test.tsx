import * as React from "react";
import * as Testing from "@testing-library/react";

import * as Util from "~/util";

import * as Component from "./foo";
import * as Styled from "./foo.styled";

// @ts-ignore
import { find } from "styled-components/test-utils";

describe("Foo", () => {
	it("renders correctly", () => {
		const node = Testing.render(<Component.Foo />);
		const FooContainer = find(node.baseElement, Styled.Foo.Container);

		expect(FooContainer).toMatchSnapshot();
	});
});
