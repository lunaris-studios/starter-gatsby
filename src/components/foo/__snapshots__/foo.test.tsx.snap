// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foo renders correctly 1`] = `
<div
  class=""
  style="opacity: 0; transform: translate3d(0%, -100%, 0%);"
  theme="[object Object]"
/>
`;
